README.md
# Treetracker Admin Panel – API

The Admin Panel is the part of the [Greenstand Treetracker project](https://github.com/Greenstand) for verifying, processing and managing data collected by the Treetracker app.

This is the RESTful API for the Admin Panel, built predominantly with [Loopback 4](https://loopback.io/doc/en/lb4/index.html).

The Admin Panel frontend is managed separately under [Greenstand/treetracker-admin-client](https://github.com/Greenstand/treetracker-admin-client).

See [Wiki](https://github.com/Greenstand/treetracker-admin-api/wiki) for more info on goals

Please add any missing content to this readme.

## Development Environment Quick Start

There are three main options for development in the Admin Panel:

1. For frontend work only
   1. Follow setup instructions in the [treetracker-admin-client](https://github.com/Greenstand/treetracker-admin-client) project
2. For API work only
   1. Fork and clone this repo as described below
   1. Use our development database credentials (available via team leads in Slack)
3. As a completely local development environment
   1. Install postgres and postgis locally, install a database seed, and run database migrations
   1. Install and run the backend API, configured to use your local database
   1. Install and run the frontend, configured to use you local backend API

### Step 1: Install git

See https://git-scm.com/downloads for instructions.

### Step 2: Install Node.js

_Node.js version 12.x works best for now; later versions have exhibited some strange behaviour with this project.
If you encounter issues with the server, check your version of Node.js first. This includes CORS related issues when fetching the API._

We recommend using [nvm](https://github.com/nvm-sh/nvm) to install and manage your Node.js instances. More details here: https://www.sitepoint.com/quick-tip-multiple-versions-node-nvm/

1. Install nvm: `curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.35.2/install.sh | bash`
2. Install the latest version of Node.js 12: `nvm install 12`
3. Use the installed Node.js: `nvm use 12`

Alternatively, you can install Node.js directly from https://nodejs.org/dist/latest-v12.x/

_On MacOS, you can alleviate the need to run as sudo by using nvm or by [following John Papa's instructions](http://jpapa.me/nomoresudo)._

### Step 3: Fork and clone this repository

1. Click _Fork_ on this GitHub repo and follow the steps to fork the repo to your account
1. Open terminal
1. Go to a folder where you would like to install the project. Then type the following, replacing `<username>` with your GitHub username:

```
git clone https://github.com/<username>/treetracker-admin-api.git
```

Add Greenstand as a remote:

```
git remote add upstream https://github.com/Greenstand/treetracker-admin-api
```

### Step 4: Get configuration files

1. Get the server dev env file pinned to the #admin_panel_chat channel in Greenstand Slack: `.env.development` (Note that the leading `.` may be removed if you download the file from Slack, so you'll need to rename it). This contains PostgreSQL development database credentials.
1. Copy the file to the root directory of your local repo

### Step 5: Install npm dependencies

```
npm install
```

### Step 6: Start the API server

```
npm start
```

### Step 7: Start developing!

## Commit Message and PR Title Format

We use automatic semantic versioning, which looks at commit messages to determine how to increment the version number for deployment.

Your commit messages will need to follow the [Conventional Commits](https://www.conventionalcommits.org/) format, for example:

```
feat: add new button
```

Since we squash commits on merging PRs into `master`, this applies to PR titles as well.

## Keeping Your Fork in Sync

Your forked repo won't automatically stay in sync with Greenstand, so you'll need to occassionally sync manually (typically before starting work on a new feature).

```
git pull upstream master --rebase
git push origin master
```

You might also need to sync and merge `master` into your feature branch before submitting a PR to resolve any conflicts.

```
git checkout <feature_branch>
git merge master
```

## Code style guide

We follow the Airbnb JavaScript style guide. The superficial aspects of this style are enforced by a pre-commit hook in the project that runs [Prettier](https://prettier.io/) when you commit a change.

If you are using VSCode as your IDE, please follow [this guide](https://www.digitalocean.com/community/tutorials/how-to-format-code-with-prettier-in-visual-studio-code) to set up Prettier and automatically format your code on file save.

You can also manually run `npm run prettier`. Configuration files are already included in this repo.

### Rules

**Indention** 2 Spaces for indentation

**Semicolon** Use semicolons at the end of each line

**Characters** 80 characters per line

**Quotes** Use single quotes unless you are writing JSON

```js
const foo = 'bar';
```

**Braces** Opening braces go on the same line as the statement

```js
if (true) {
  console.log('here');
}
```

**Variable declaration** Declare one Variable per statement

```js
const dog = ['bark', 'woof'];
let cat = ['meow', 'sleep'];
```

**Variable, properties and function names** Use lowerCamelCase for variables, properties and function names

```js
const adminUser = db.query('SELECT * From users ...');
```

**Class names** Use UpperCamelCase for class names

```js
class Dog {
  bark() {
    console.log('woof');
  }
}
```

**Descriptive conditions** Make sure to have a descriptive name that tells the use and meaning of the code

```js
const isValidPassword =
  password.length >= 4 && /^(?=.*\d).{4,}$/.test(password);
```

**Object/Array creation** Use trailing commas and put short declarations on a single line. Only quote keys when your interpreter complains:

```js
var a = ['hello', 'world'];
var b = {
  good: 'code',
  'is generally': 'pretty',
};
```

## Testing

We used a combination of JS and Typescript, and because Loopback would load services/controllers from the typescript output folder (dist), it can be tricky to test.

For the goal of protecting the shared development database, when running test, we will use a separate database.

Create a test environment file `.env.test` in the root directory with the test database URL set as follows:

```
DATABASE_URL=postgresql://<user>:<password>@<host>:<port>/<database>?ssl=true
```

NOTE: Please do not set this URL to point to our development database, because the tests will clear all the data in the database. It would cause trouble if we don't have any data in the dev DB.

To locally install postgresDB, this app might be helpful: https://postgresapp.com/

To run test:

```
npm test
```

To make the test process more smooth, we suggest running a command to compile the Loopback files automatically when files change:

```
npm run watch
```

In this way, we can write the code and get the tests result immediately.

NOTE: when running tests, the files related to Loopback are loaded from ./dist folder. That's because for Jest does not output compiled files at all, and Loopback will try to load the controllers at runtime.

## Advanced local development using docker _## Currently broken ##_

For developers familiar with docker, we offer a dockerized setup for local development.

To run docker on a local machine, you will have to install Docker first.
Docker is a linux container technology, so running it on Mac or Windows requires an application with an attached linux VM.
Docker provides one for each OS by default.

### Mac

Install Docker for Mac using homebrew, using the following command

```
$ brew cask install docker
```

You can alternatively install Docker via: [Docker for Mac](https://docs.docker.com/docker-for-mac/install/)

Once Docker is installed, lauch Docker from the Applications GUI.

### Windows

For most versions of Windows: [Docker for Windows](https://docs.docker.com/docker-for-windows/install/)

For some older versions or Win10 Home: [Docker Toolbox](https://docs.docker.com/toolbox/toolbox_install_windows/).
At least on one machine, to get this to work, when you get to the step to do QuickStart terminal script, instead, run:

```
docker-machine create default --virtualbox-no-vtx-check
```

then re-run the QuickStart terminal script.

> If you use Docker Toolbox, check the IP address in the output of the QuickStart terminal script.
> You will use this IP address later instead of `localhost`.

### Linux

To install on linux, you can run
`sudo apt-get install -y docker-ce`
but there is [additional setup](https://docs.docker.com/install/linux/docker-ce/ubuntu/#set-up-the-repository) to verify keys, etc.

### Install, build docker containers and go

Run the setup script. This script installs node modules, builds docker containers, and starts them

```
./dev/scripts/setup.sh
```

You can now view the Treetracker Admin Panel at http://localhost:8080.

> Note: If you try to access the site on port 3001 you will recieve a CORS error

> Note: If you used Docker Toolbox, you may need to use the IP address it reported, such as http://192.168.99.100:8080_

It may take a few seconds for the web and api servers to come up. You can monitor them using the docker logs commands as:

```
docker logs -f treetracker-admin-web
docker logs -f treetracker-admin-api
```

Also see [Scripts](#scripts) below

To stop the dev environment use

```
./dev/scripts/down.sh
```

To start the dev environment back up use

```
./dev/scripts/up.sh
```

Just edit as you normally would to view changes in your development environment.

### Alternative setup for MS Windows (Works on Linux and Mac also)

On Windows the easiest way to develop and debug Node.js applications is using Visual Studio Code.
It comes with Node.js support out of the box.

https://code.visualstudio.com/docs

### Still can not figure it out?

Here is our [wiki page for troubleshooting](https://github.com/Greenstand/treetracker-admin/wiki/Set-Up-Issues), take a look.

Help us to improve it by adding your experience solving this problem.

### Scripts

Useful scripts are contained in /dev/scripts. Their uses are described here. Scripts are run from the repository root as /dev/scripts/{script-name}.sh

**install.sh** install or update npm modules for server and client projects

**build.sh** build docker images

**up.sh** bring up docker containers in docker as described by docker-compose.yml

**setup.sh** run install.sh, build.sh, and up.sh

**down.sh** bring down docker containers

**logs-api.sh** show logs for api server

**logs-web.sh** show logs for React.js dev server

**docker-clear-images.sh** clear out _all_ docker images

**docker-remove-containers.sh** clear out _all_ docker containers

## Further reading

See [Contributing to the Cause](https://github.com/Greenstand/Development-Overview#contributing-to-the-cause)


CHANGELOG.md
## [2.26.3](https://github.com/Greenstand/treetracker-admin/compare/v2.26.2...v2.26.3) (2022-11-01)


### Bug Fixes

* allow DB to create species UUID ([0322686](https://github.com/Greenstand/treetracker-admin/commit/03226868660d49cd2b76191340557050c713fd87))

## [2.26.2](https://github.com/Greenstand/treetracker-admin/compare/v2.26.1...v2.26.2) (2022-11-01)


### Bug Fixes

* allow deployment unless commit message starts with chore ([c358530](https://github.com/Greenstand/treetracker-admin/commit/c3585301fbbab5d38bc4372f204a17bc84a71c8c))

## [2.26.1](https://github.com/Greenstand/treetracker-admin/compare/v2.26.0...v2.26.1) (2022-11-01)


### Bug Fixes

* create UUID for new and combined species ([#651](https://github.com/Greenstand/treetracker-admin/issues/651)) ([04626f6](https://github.com/Greenstand/treetracker-admin/commit/04626f653d79e178a5335f767cdcfba30581e6d5))

# [2.26.0](https://github.com/Greenstand/treetracker-admin/compare/v2.25.5...v2.26.0) (2022-10-07)


### Features

* add uuid to species and tag models ([#647](https://github.com/Greenstand/treetracker-admin/issues/647)) ([91fffae](https://github.com/Greenstand/treetracker-admin/commit/91fffaebafb3cf732e9ec2b82045f8ca2bdbf8c4))

## [2.25.5](https://github.com/Greenstand/treetracker-admin/compare/v2.25.4...v2.25.5) (2022-06-16)


### Bug Fixes

* updated npm packages that had security issues ([#642](https://github.com/Greenstand/treetracker-admin/issues/642)) ([5ff9f0c](https://github.com/Greenstand/treetracker-admin/commit/5ff9f0ca274916755f5d4140b6dbf80e89e0a0d0))

## [2.25.4](https://github.com/Greenstand/treetracker-admin/compare/v2.25.3...v2.25.4) (2022-06-15)


### Bug Fixes

* remove duplicates when filtering with tags ([#641](https://github.com/Greenstand/treetracker-admin/issues/641)) ([cbc6c46](https://github.com/Greenstand/treetracker-admin/commit/cbc6c462a0314b8f76aeadf4310b72f4a4c93c37))

## [2.25.3](https://github.com/Greenstand/treetracker-admin/compare/v2.25.2...v2.25.3) (2022-03-12)


### Bug Fixes

* return planter registrations where device identifier is null ([#625](https://github.com/Greenstand/treetracker-admin/issues/625)) ([5fbb81a](https://github.com/Greenstand/treetracker-admin/commit/5fbb81a8ab3e8ca4a8076f06d656d8b7fbbba5d0))

## [2.25.2](https://github.com/Greenstand/treetracker-admin/compare/v2.25.1...v2.25.2) (2022-03-07)


### Bug Fixes

* hide grower_account_id field from planet until api is planned ([#623](https://github.com/Greenstand/treetracker-admin/issues/623)) ([1c90e2b](https://github.com/Greenstand/treetracker-admin/commit/1c90e2b14f7d949801ab654cc7c571f0d9a6227f))

## [2.25.1](https://github.com/Greenstand/treetracker-admin/compare/v2.25.0...v2.25.1) (2022-03-05)


### Bug Fixes

* disable API timeout ([#622](https://github.com/Greenstand/treetracker-admin/issues/622)) ([b1c2173](https://github.com/Greenstand/treetracker-admin/commit/b1c2173fb08bec8ecb31e65bbdff7c84c928f024))

# [2.25.0](https://github.com/Greenstand/treetracker-admin/compare/v2.24.2...v2.25.0) (2022-02-11)


### Features

* add stakeholder-uuid ([38d5010](https://github.com/Greenstand/treetracker-admin/commit/38d5010f68deb1b82574ae33413556191aa50b0b))

## [2.24.2](https://github.com/Greenstand/treetracker-admin/compare/v2.24.1...v2.24.2) (2022-01-26)


### Bug Fixes

* allow manager_user and list_user policies to access user manager ([#614](https://github.com/Greenstand/treetracker-admin/issues/614)) ([0d23f81](https://github.com/Greenstand/treetracker-admin/commit/0d23f81a623c54c817292c1b31db199698fd0fb0))

## [2.24.1](https://github.com/Greenstand/treetracker-admin/compare/v2.24.0...v2.24.1) (2022-01-16)


### Bug Fixes

* don't change org type for children ([840e6b9](https://github.com/Greenstand/treetracker-admin/commit/840e6b997031a27f8416f4f72f56a34cbe7f6e95))

# [2.24.0](https://github.com/Greenstand/treetracker-admin/compare/v2.23.3...v2.24.0) (2022-01-10)


### Features

* add device manufacturer to planter registration data for both admin and org accounts ([74c919b](https://github.com/Greenstand/treetracker-admin/commit/74c919ba9c8a8bb9077fed70ea2fbb05e778d4a9))

## [2.23.3](https://github.com/Greenstand/treetracker-admin/compare/v2.23.2...v2.23.3) (2022-01-04)


### Reverts

* Revert "fix: add DISTINCT clause to ensure no duplicates (#611)" (#612) ([623d1d0](https://github.com/Greenstand/treetracker-admin/commit/623d1d0cc7451dd6a94ba895e8c582a305bd5605)), closes [#611](https://github.com/Greenstand/treetracker-admin/issues/611) [#612](https://github.com/Greenstand/treetracker-admin/issues/612)

## [2.23.2](https://github.com/Greenstand/treetracker-admin/compare/v2.23.1...v2.23.2) (2022-01-04)


### Bug Fixes

* add DISTINCT clause to ensure no duplicates ([#611](https://github.com/Greenstand/treetracker-admin/issues/611)) ([464619c](https://github.com/Greenstand/treetracker-admin/commit/464619c74e0e0371d5c2288ab5e685bccf038c9a))

## [2.23.1](https://github.com/Greenstand/treetracker-admin/compare/v2.23.0...v2.23.1) (2021-12-17)


### Bug Fixes

* allow CORS from staging ([2c15ee2](https://github.com/Greenstand/treetracker-admin/commit/2c15ee2580837d9da6d27aea306d9c19c9eab417))

# [2.23.0](https://github.com/Greenstand/treetracker-admin/compare/v2.22.2...v2.23.0) (2021-12-01)


### Features

* add join clause for trees with any tagid ([#608](https://github.com/Greenstand/treetracker-admin/issues/608)) ([c436501](https://github.com/Greenstand/treetracker-admin/commit/c4365016924783fdee1b92939bd0c7a2f05a2adb))

## [2.22.2](https://github.com/Greenstand/treetracker-admin/compare/v2.22.1...v2.22.2) (2021-11-23)


### Bug Fixes

* **planter-registration-repo:** change selector back to * so that country info would be retrieved ([#605](https://github.com/Greenstand/treetracker-admin/issues/605)) ([2d612a9](https://github.com/Greenstand/treetracker-admin/commit/2d612a9813ef4e0dc08c59f67afca13361b4181a))

## [2.22.1](https://github.com/Greenstand/treetracker-admin/compare/v2.22.0...v2.22.1) (2021-11-22)



index.js
var express = require('express');
var bodyParser = require('body-parser');
var http = require('http');
var pg = require('pg');
const { Pool, Client } = require('pg');
var path = require('path');
var app = express();
var port = process.env.NODE_PORT || 3000;
var conn = require('./config');

var async = require("async");

var moment = require("moment");


app.use(bodyParser.urlencoded({ extended: false })); // parse application/x-www-form-urlencoded
app.use(bodyParser.json()); // parse application/json
app.set('view engine', 'html');


const noop = function () { }

const isNotEmpty = function (val) {
  return (val !== null && val !== undefined);
}

//2-15-2018 pool deprecated for this custom module, use crudpool instead
//const pool = new Pool({

//connectionString: conn.connectionString
//});

const crudpool = new Pool({

  connectionString: conn.connectionString
});


const shouldabort = (err, client) => {
  if (err) {
    console.error('Error in transaction', err.stack);
    client.query('ROLLBACK', (err2) => {
      if (err2) {
        console.error('Error rolling back client', err2.stack);
      }
    });
  }
  return !!err;
}





const connectpool = function (pool, cb) {
  pool.connect(function (err, client, done) {
    if (err) {
      console.error('Error in transaction', err.stack);
    }
    else {

      cb(client, done);
    }

  });
};


const beginquery = function (client, cb) {
  client.query('BEGIN', (err1) => {
    if (shouldabort(err1, client)) return;
    cb();
  })

};

const issuequery = function (client, querytext, queryvalues, cb) {
  client.query(querytext, queryvalues, (err2, res) => {
    if (shouldabort(err2, client)) return;
    cb(res);
  })

};


const commitquery = function (client, cb) {
  client.query('COMMIT', (err) => {
    if (err) {
      console.error('Error committing transaction', err.stack);
    }
    else {
      cb();
    }

  })

};

const getUpdateTableQueryByTableIDCols = function (table, id, cols, cbb) {
  // Setup static beginning of query with the table too
  var query = ['UPDATE ' + table];
  query.push('SET');

  // Create another new array storing each set command there
  // and then assigning a number value for parameterized query there
  var mySet = [];


  const f1 = function (itm, idx, cb) {
    mySet.push(itm + ' = ($' + (idx + 1) + ')');
    cb();
  };

  const f2 = function (err1) {
    if (err1) {
      console.error('Error async eachSeries loop 2_f2', err1.stack)
    }
    else {
      query.push(mySet.join(', '));
      query.push('WHERE id = ' + id);
      cbb(query.join(' '));
    }
  };

  async.eachOfSeries(Object.getOwnPropertyNames(cols), f1, f2);
}

const getFormattedUTCTimeNow = function () {

  return moment.utc().format("YYYY-MM-DD HH:mm:ss+00");

}





//start with connectpool to get a client only once.
// Keep this client open and reuse for atomicity
//done releases the client. don't call done(); Keep reusing the same client instead
//pg isolates transaction by client. Calling new client defeats the atomicity
//and the purpose of BEGIN and COMMIT
//So reuse same client -> crudclient

///////////////////

var crudclient = null;
var crud_done = null;

connectpool(crudpool, function (client, done) {
  crudclient = client;
  crud_done = done;
});

//////////////////

//////////////////


/*
app.post('/trees/create', function(req, res){
  var user_id = req.body["user_id"];
  var lat = req.body["lat"];
  var lon = req.body["lon"];
  var gps_accuracy = req.body["gps_accuracy"];
  var note = req.body["note"];
  var timestamp = req.body["timestamp"];
  var base_64_image = req.body["base_64_queryimage"];

//console.log("timestamp"); 
//console.log(timestamp);
//console.log(req.body);
//console.log(req.body["timestamp"]);

        const insertTreeText= 'INSERT INTO trees(user_id, lat, lon, gps_accuracy, note, timestamp, base_64_image) VALUES ($1, $2, $3, $4, $5, $6, $7)';
        const insertTreeValues = [user_id, lat, lon, gps_accuracy, note, timestamp, base_64_image];


        beginquery(crudclient, function(){
           issuequery(crudclient,insertTreeText,insertTreeValues,function(res1){
              commitquery(crudclient,function(){

              });
           });
        });


});
*/

app.delete('/trees', function (req, res) {
  var id = req.body["id"];
  const deleteTreeText = "DELETE FROM trees WHERE id = $1";
  const deleteTreeValues = [id];

  beginquery(crudclient, function () {
    issuequery(crudclient, deleteTreeText, deleteTreeValues, function (res1) {
      commitquery(crudclient, function () {
        res.status(200).json({
          data: res1
        });
      });
    });
  });


});


app.put('/trees', function (req, res) {
  var id = req.body["id"];
  //var time_created = req.body["time_created"];
  //var time_updated = req.body["time_updated"];
  var missing = req.body["missing"];
  var priority = req.body["priority"];
  var cause_of_death_id = req.body["cause_of_death_id"];
  var user_id = req.body["user_id"];
  var primary_location_id = req.body["primary_location_id"];
  var settings_id = req.body["settings_id"];
  var override_settings_id = req.body["override_settings_id"];
  var dead = req.body["dead"];
  var lat = req.body["lat"];
  var lon = req.body["lon"];
  //var estimated_geometric_location = req.body["estimated_geometric_location"];
  //var gps_accuracy = req.body["gps_accuracy"];

  var time_updated = getFormattedUTCTimeNow();
  //console.log("time_updated");
  //console.log(time_updated);

  var treecols =
    {
      "time_updated": time_updated

    }

  isNotEmpty(missing) ? treecols["missing"] = missing : noop;
  isNotEmpty(priority) ? treecols["priority"] = priority : noop;
  isNotEmpty(cause_of_death_id) ? treecols["cause_of_death_id"] = cause_of_death_id : noop;
  isNotEmpty(user_id) ? treecols["user_id"] = user_id : noop;
  isNotEmpty(primary_location_id) ? treecols["primary_location_id"] = primary_location_id : noop;
  isNotEmpty(settings_id) ? treecols["settings_id"] = settings_id : noop;
  isNotEmpty(override_settings_id) ? treecols["override_settings_id"] = override_settings_id : noop;
  isNotEmpty(dead) ? treecols["dead"] = dead : noop;
  isNotEmpty(lat) ? treecols["lat"] = lat : noop;
  isNotEmpty(lon) ? treecols["lon"] = lon : noop;


  const fn5 = function (updateTreeText) {



    var updateTreeValues = [];

    //prefer Object.getOwnPropertyNames() over Object.keys() because:
    //Object.getOwnPropertyNames() is more likely to guarantee that 
    //JS-Object to ArrayofJS-Object_AllKeys conversion
    //preserves JS Object keys creation order in the returned array

    const updateTreeKeys = Object.getOwnPropertyNames(treecols);



    //used in async.eachSeries call, the following function is the callback function
    //the params are (some_custom_error) and it should usually be blank
    //it is called with signature callback() when all elements finished iterating,
    //or if "callback(some_custom_error)" was manually called 
    //from the iteratee during iteration      

    //BEGINNING of callback function
    const updateTreeValuesReady = function (err) {
      if (err) {
        console.error('Error async eachSeries loop 1', err.stack)
      }
      else {
        //console.log('updateTreeText');
        //console.log(updateTreeText);
        //console.log('updateTreeValues');
        //console.log(updateTreeValues);


        beginquery(crudclient, function () {
          issuequery(crudclient, updateTreeText, updateTreeValues, function (res1) {
            commitquery(crudclient, function () {
              res.status(200).json({
                data: res1
              });
            });
          });
        });
      }

    }

    //END of callback function



    //used in async.eachSeries call, the following function is the iteratee function
    //the iteratee function signature is (item, callback)
    //see EACHSERIES FUNCTION CALL for more details  
    //the async utility automatically populates "item" with the current item from collection
    //and also automatically populates "callback" with something that is explained more
    //in EACHSERIES FUNCTION CALL section     

    //BEGINNING of iteratee function 
    const turnTreecolsIntoArrayofValues = function (item, callback) {
      var treeValue = treecols[item];
      updateTreeValues.push(treeValue);
      callback();

    }

    //END of iteratee function



    //The following turns "treecols" JS Object into an array of values for updateTreeValues

    //EACHSERIES FUNCTION CALL
    //Using async.eachSeries loops through each array (collection) item
    //asynchronously yet sequentially.
    //the params are (collection,iteratee,callback) and all three should be included
    //collection is usually an array, iteratee and callback are usually functions
    //iteratee function signature is (item, callback)
    //callback function signature is (err) and err should be usually blank
    //in this example, calling the "callback" parameter from within the iteratee function 
    //does one of three things:
    //1. If there are any remaining elements in the passed in array (collection)
    //and callback is called like this: callback();
    //it repeats this same iteratee function again 
    //but this time, the "item" param contains the next item from the passed in array.        
    //2. if there are no more items in array (collection)
    //and callback is called like this: callback();
    //the actual callback function is called, with the error parameter empty.
    //3. Sometimes, even if there are still remaining items in the array (collection),
    //but if anytime during iteration, this iteratee function calls 
    //callback(some_custom_error) instead of just callback();
    //then async utility jumps out of the iteratee immediately in this case, 
    //and the callback is called immediately with the error parameter populated
    //with some_custom_error
    //
    async.eachSeries(updateTreeKeys, turnTreecolsIntoArrayofValues, updateTreeValuesReady);
  }

  getUpdateTableQueryByTableIDCols("trees", id, treecols, fn5);


});


app.get('/trees', function (req, res) {
  console.log(req.url);

  var sql = "SELECT 'point' AS type, trees.* FROM trees";
  var values = []
  var query = {}



  if (req.query['zoom'] < 14) {
    let zoom = req.query['zoom'];
    sql = "SELECT 'cluster' AS type, ST_AsGeoJSON(ST_Centroid(clustered_locations)) centroid, ST_AsGeoJSON(ST_MinimumBoundingCircle(clustered_locations)) circle, ST_NumGeometries(clustered_locations) count FROM ( SELECT unnest(ST_ClusterWithin(estimated_geometric_location, $1)) clustered_locations from trees ) clusters";
    values.push(zoom)

  }

  if (req.query['user_id']) {
    let user_id = req.query['user_id']
    sql = sql + " WHERE user_id = " + user_id
  }

  if (req.query['start_index']) {
    let start_index = parseInt(req.query['start_index'])
    sql = sql + " OFFSET " + (start_index - 1)
  }

  if (req.query['page_size']) {
    let page_size = req.query['page_size']
    sql = sql + " LIMIT " + page_size
  }

  query = {
    text: sql,
    values: values
  }
  //console.log()
  ///////////////////////////////////////////
  //
  ////ok for now and in this case only, 
  ////but do not use pool.query in general - 
  ////prefer to use client.query - 1-23-2018
  ////pool.query(query)
  ////.then(function(data){
  ////res.status(200).json({              
  ////data: data.rows
  ////})
  ////})
  ////.catch(e => console.error(e.stack));
  ////////////////////////////////////////////////

  //2-15-2018 using crudpool convention even for reading data

  beginquery(crudclient, function () {
    issuequery(crudclient, query.text, query.values, function (res1) {
      commitquery(crudclient, function () {
        res.status(200).json({
          data: res1.rows
        });
      });
    });
  });

});

app.listen(port, () => {
  console.log('listening on port ' + port);
});
